docs/proposals/backend-support-appset.md

```diff
@@ -0,0 +1,226 @@
+Example: if you try to delete an ApplicationSet, we check if you can delete it’s child Applications.
+
+#### **ApplicationSet Deletion Algorithm:**
+
+A user is only able to delete an ApplicationSet if they have permissions to delete all of the Applications managed by the ApplicationSet. This check is performed in ApplicationSet controller, on receiving a delete request via GRPC from API server.
+
+* For each application owned by the ApplicationSet that the user is attempting to delete:
+    * Check if the user has delete permission on the Application
+    * Check if the user has delete permission within the project (?)
+* If the user does NOT have permission on least one of these, the operation should fail.
+* On pass, ApplicationController server deletes (ie kubectl delete) the ApplicationSet resource.
+
+#### **ApplicationSet Update Algorithm:**
+
+A user can only update an ApplicationSet if the user has permission to update all of the Applications currently owned by the ApplicationSet.
+
+* When the user makes a change to an ApplicationSet, we assume that it's possible that the change might affect any or all of the Applications, and thus we require the user to have write access to all of those Applications.
+* We likewise check that the resulting generated Applications are also compliant with the user's permissions.
+
+Algorithm is, if the user attempts to update an ApplicationSet via Web UI/CLI:
+
+* ApplicationSet controller receives a request to update an ApplicationSet from API server
+* The ApplicationSetController looks at all the Applications owned by the ApplicationSet (via ownerref or annotation):
+    Verify that the user has permission to act on all of the Applications currently managed by the ApplicationSet
+* If the above precondition is met, proceed to the next step, otherwise fail.
+* The ApplicationSet is generated and rendered into a template
+  * Look at the generated Applications, and make sure the user has appropriate privileges
+    * All the same checks done by the Create workflow, described above, are done here (user can access repo, cluster, etc
+* Finally, on success, the API server applies (kubectl apply) the requested change to the ApplicationSet (and the Applications).
+
+#### **ApplicationSets and AppProjects:**
+
+An important design constraint in this area: ApplicationSets do not belong to projects. They generate Applications that are a part of projects, but they themselves are not part of a project.
+
+1. Why not just include a project field on ApplicationSet CR?
+
+   A single ApplicationSet has the power to produce Applications within multiple projects, so it does not necessarily make sense to include an ApplicationSet within a single project.
+
+2. Why not just include a projects field (array of strings) on ApplicationSets, to allow it to belong to multiple projects?
```

Comment:
I think this is a good idea, except instead of "belonging" to multiple projects, the appset should have a field to "allow-list" certain projects. If the generated application's project is not on the allow-list, the appset controller should refuse to create it. The allow-list would allow the user to create appsets which are constrained to the user's permissions.

Reply:
That sounds like a good idea. @jgwest @chetan-rns WDYT?

Reply:
Same issue as described in the answer: providing a hardcoded, statically defined list of projects potentially limits the expressive power of the ApplicationSet.

However, [as was brought up elsewhere](https://github.com/argoproj/argo-cd/pull/9016/files#r846005099), since ApplicationSets are so (expressively) powerful, it's very difficult to determine what one will do at creation, so tough (but necessary) to ensure security bounds. Thus it may be a necessary evil to add this 'list of projects' restriction :smile:.

Reply:
I think if it's optional, it'll be alright. Users with `*` project access can still create whatever appsets they want, but users with access to just a few projects wouldn't be limited to hard-coding one project in their appset - they could make the field dynamic and restrict themselves with an allowlist.